### ISSUE TITLE
Feedback wanted: Support for components added using Livewire

### ISSUE THREAD

**Issue by stancl** [alpinejs/alpine#1144]
Right now, Alpine only works with Livewire components that were present on the initial load. Alpine components added using Livewire components which were requested later seem to **initialize before Livewire**, which makes the entire JS runtime break and stop executing.

I found a way to solve this, by essentially blocking Alpine component updates/inits while Livewire does its job. I think this is the right approach.

However, a few changes are needed to both project's JS suite, and I have no idea how to write tests for this, so I'm submitting this as an issue.

### Waiting for Livewire

```diff
updateElements(rootEl, extraVars = () => {}) {
+   let update = () => {
        this.walkAndSkipNestedComponents(rootEl, el => {
            // Don't touch spawns from for loop (and check if the root is actually a for loop in a parent, don't skip it.)
            if (el.__x_for_key !== undefined && ! el.isSameNode(this.$el)) return false
    
            this.updateElement(el, extraVars)
        }, el => {
            el.__x = new Component(el)
        })
    
        this.executeAndClearRemainingShowDirectiveStack()
    
        this.executeAndClearNextTickStack(rootEl)
+   }

+    if (window.LivewireManagingDOM) {
+        window.LivewireManagingDOM.then(update)
+    } else {
+        update();
+    }
}

initializeElements(rootEl, extraVars = () => {}, shouldRegisterListeners = true) {
+    let initialize = () => {
        this.walkAndSkipNestedComponents(rootEl, el => {
                // Don't touch spawns from for loop
                if (el.__x_for_key !== undefined) return false; // Don't touch spawns from if directives
  
                if (el.__x_inserted_me !== undefined) return false;
  
                this.initializeElement(el, extraVars, shouldRegisterListeners);
            }, el => {
                el.__x = new Component(el);
            });
            this.executeAndClearRemainingShowDirectiveStack();
            this.executeAndClearNextTickStack(rootEl);
+    }

+    if (window.LivewireManagingDOM) {
+        window.LivewireManagingDOM.then(initialize)
+    } else {
+        initialize()
+    }
}
```

### Skipping transitions

To support `$wire` updates etc, Livewire runs Alpine's `updateElements()` after it finishes making changes to the DOM.

This results in `show="false"` transitions thinking that they were user-initiated, so as the element is removed from the DOM, it runs `transitionOut()`.

```diff
function transitionIn(el, show, reject, component, forceSkip = false) {
+    if (window.skipAlpineTransitions) return show();

function transitionOut(el, hide, reject, component, forceSkip = false) {
+    if (window.skipAlpineTransitions) return hide();
```

## Livewire changes

The `refreshAlpineAfterEveryLivewireRequest()` method should be replaced with logic that works like this:

```js
window.Livewire.hook('message.sent', () => {
    window.LivewireManagingDOM = new Promise((resolve, reject) => {
        window.Livewire.hook('message.processed', () => {
            window.LivewireManagingDOM = undefined;

            resolve(true);
        });

        window.Livewire.hook('message.failed', () => {
            window.LivewireManagingDOM = undefined;

            reject();
        });
    }).then(() => {
        window.skipAlpineTransitions = true;
        window.LivewireManagedAlpineComponents.forEach(component => {
            component.updateElements(component.$el);
        });
        window.skipAlpineTransitions = false;
    });
});

window.LivewireManagedAlpineComponents = [];

window.Alpine.onComponentInitialized(component => {
    let livewireEl = component.$el.closest('[wire\\:id]')

    if (livewireEl && livewireEl.__livewire) {
        window.LivewireManagedAlpineComponents.push(component);
    }
})
```

This creates a Promise when a LW message is sent, and resolves it when a response is **fully processed**. The Alpine logic was moved to `.then()`, meaning that **Alpine will only execute after LW is done with its changes**.

It also runs the original `refreshAlpineAfterEveryLivewireRequest()` logic, but it does so *after* the components have been properly initialized first.

## More context

The variable names used in this code are obviously POC. I think it'd make sense to make this non-LW specific in Alpine, and then hook into that logic from LW.

And the `window.skipAlpineTransitions` should probably be a property on `window.Alpine`.

cc @joshhanley

**Comment by stancl** [comment]
I tried to recreate this a few more times and seems like the transition changes may not be necessary, and can be avoided by structuring your code in a bit different way.

So that's not *as* important, and I think it should first be reproduced before an attempt at fixing it is made.

**Comment by stancl** [comment]
@calebporzio FWIW Josh's patch didn't fix all of the problems that I encountered, and overall Alpine seemed to work the best when I combined Josh's changes with mine.

It's somewhat difficult to reproduce this though, and I managed to patch this at runtime myself. Not ideal, but it works well enough.

If someone reports this again and can give a reliable example that reproduces this, it may be worth revisiting the changes proposed here.

Otherwise it's probably not too high priority
